test/parallel/test-headers.js

```diff
@@ -0,0 +1,491 @@
+'use strict';
+
+require('../common');
+const assert = require('assert');
+
+// Flags: --expose-internals
+
+const {
+  Headers,
+  kHeadersList,
+  binarySearch
+} = require('internal/fetch/headers');
+
+{
+  // init is undefined
+  assert.deepStrictEqual(new Headers()[kHeadersList], []);
+  // Init is flat array with one entry
+  assert.deepStrictEqual(
+    new Headers(['test-name', 'test-value'])[kHeadersList],
+    ['test-name', 'test-value']
+  );
+  // Init is flat array with multiple entries
+  assert.deepStrictEqual(
+    new Headers(['test-name-1', 'test-value-1', 'test-name-2', 'test-value-2'])[
+      kHeadersList
+    ],
+    ['test-name-1', 'test-value-1', 'test-name-2', 'test-value-2']
+  );
+  // Init is multidimensional array with one entry
+  assert.deepStrictEqual(
+    new Headers([['test-name-1', 'test-value-1']])[kHeadersList],
+    ['test-name-1', 'test-value-1']
+  );
+  // Init is multidimensional array with multiple entries
+  assert.deepStrictEqual(
+    new Headers([
+      ['test-name-1', 'test-value-1'],
+      ['test-name-2', 'test-value-2'],
+    ])[kHeadersList],
+    ['test-name-1', 'test-value-1', 'test-name-2', 'test-value-2']
+  );
+  // Throws when init length is odd
+  assert.throws(() => {
+    new Headers(['test-name-1', 'test-value', 'test-name-2']);
+  }, {
+    name: 'TypeError',
+    message: "The argument 'init' is not even in length. " +
+      "Received [ 'test-name-1', 'test-value', 'test-name-2' ]"
+  });
+  // Throws when multidimensional init entry length is not 2
+  assert.throws(() => {
+    new Headers([['test-name-1', 'test-value-1'], ['test-name-2']]);
+  }, {
+    name: 'TypeError',
+    message: "The argument 'init' is not of length 2. " +
+      "Received [ 'test-name-2' ]"
+  });
+  // Throws when init is not valid array input
+  assert.throws(() => {
+    new Headers([0, 1]);
+  }, {
+    name: 'TypeError',
+    message: "The argument 'init' is not a valid array entry. " +
+      'Received [ 0, 1 ]'
+  });
+}
+
+{
+  // Init is object with single entry
+  const headers = new Headers({
+    'test-name-1': 'test-value-1'
+  });
+  assert.strictEqual(headers[kHeadersList].length, 2);
+}
+
+{
+  // Init is object with multiple entries
+  const headers = new Headers({
+    'test-name-1': 'test-value-1',
+    'test-name-2': 'test-value-2'
+  });
+  assert.strictEqual(headers[kHeadersList].length, 4);
+}
+
+{
+  // Init fails silently when initialized with BoxedPrimitives
+  [
+    new Number(),
+    new Boolean(),
+    new String(),
+  ].forEach((arg) => new Headers(arg));
+}
+
+{
+  // Init fails silently if function or primitive is passed
+  [
+    new Function(),
+    function() {},
+    1,
+    'test',
+  ].forEach((arg) => new Headers(arg));
```

Comment:
This test case is failing cause of this line:

```js
validateObject(init, 'init', { allowArray: true }) === undefined)
```

Any ideas how to fix this?

Reply:
Do we have something like "isPrimitive"?

Reply:
Resolving for a typeof check for now. welcome to other ideas

Reply:
A better resolution would be to not allow primitives, like browsers do IMO.

Reply:
removed this so it follows spec